---
title: "cis-meQTLs Analysis. meSNP Enrichment Analysis"
subtitle: "Dexamethasone-Stimulated Human Array Project"
author: 
  - name: Anastasiia Hryhorzhevska
    email: anastasiia_hry@psych.mpg.de
    url: https://github.com/ahryho
    affiliation: Max Planck Institute of Psychiatry
    affiliation_url: https://www.psych.mpg.de/2664393/medizinische-genomforschung
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    css: ../style.css
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, include=FALSE}
options(digits = 4, width = 100, stringsAsFactors = T)
knitr::opts_chunk$set(echo = TRUE,
                      tidy.opts = list(width.cutoff = 100),
                      tidy=TRUE,
                      fig.pos = "H",
                      dpi = 200,
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE, 
                      cache.lazy = FALSE)

# define DT as the standard printing method for data.frame
library(knitr)
library(DT)
knit_print.data.frame <- function(x, ...) {
  knit_print(DT::datatable(x,
                           filter="top",
                           rownames = FALSE,
                           extensions = "FixedColumns",
                           options = list(
                             scrollX = TRUE,
                             pageLength = 5
                             
                           )), ...)
}

registerS3method("knit_print", "data.frame", knit_print.data.frame)
```

# __Functional SNP annotation__
***

```{r, include = F, eval = T}
source("~/bio/code/mpip/dex-stim-human-array/code/util.R", chdir = TRUE)

library(data.table)
library(dplyr)

library(ggplot2)
library(ggrepel)
library(scales)
library(GGally)
library(factoextra)
library(viridis)

library(splitstackshape)
library(reshape2)
library(IRanges)
library(biomaRt)

require(foreign)
library(parallel)
library(foreach)
library(doParallel)

library(gUtils)

out.dir.pre      <- "/binder/mgp/workspace/2020_DexStim_Array_Human/output/data/integrative/matrixEQTL/meqtls/region_wise_independent_snps/"
out.dir.pre.pre  <- "/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-array/output/data/integrative/matrixEQTL/meqtls/"

# meqtl.dex.fn   <- paste0(out.dir.pre, "me-qtl_cis_indp_result_dex_with_delta_fdr_005.csv")
# meqtl.veh.fn   <- paste0(out.dir.pre, "me-qtl_cis_indp_result_veh_fdr_005.csv")
# meqtl.delta.fn <- paste0(out.dir.pre, "me-qtl_cis_indp_result_delta_fdr_005.csv")

# col.names <-  c("SNP", "CpG_ID", "beta", "t-stat", "p-value", "fdr")

# ind.meqtl.dex.df   <- fread(meqtl.dex.fn, col.names = col.names)
# ind.meqtl.veh.df   <- fread(meqtl.veh.fn, col.names = col.names)
# ind.meqtl.delta.df <- fread(meqtl.delta.fn, col.names = col.names)
```

```{r include = F, eval = T}
# code: ../code/methylation/02_dma/01_dmp/03_get_cpg_gene_anno_dist
# cpg.closest.genes.dist.fn <-"~/bio/code/mpip/dex-stim-human-array/output/data/methylation/02_dmp/dex_cpgs_annotated_closest_genes_distances.csv"
# cpg.closest.genes.dist.df <- fread(cpg.closest.genes.dist.fn)

delta.meqtl.snp.anno.rds              <- readRDS(paste0(out.dir.pre, "meqtls_snp_annotated_withChIPseeker_delta.rds"))
# delta.meqtl.cpg.closest.genes.dist.df <- cpg.closest.genes.dist.df[PROBE_ID %in% delta.meqtl.cpg.anno.df$Name]

veh.meqtl.snp.anno.rds                <- readRDS(paste0(out.dir.pre, "meqtls_snp_annotated_withChIPseeker_veh.rds"))
# veh.meqtl.cpg.closest.genes.dist.df   <- cpg.closest.genes.dist.df[PROBE_ID %in% veh.meqtl.cpg.anno.df$Name]

dex.meqtl.snp.anno.rds                <- readRDS(paste0(out.dir.pre, "meqtls_snp_annotated_withChIPseeker_dex.rds"))
# dex.meqtl.cpg.closest.genes.dist.df   <- cpg.closest.genes.dist.df[PROBE_ID %in% dex.meqtl.cpg.anno.df$Name]
```

## Illumina {.tabset .tabset-fade .tabset-pills}

### Example of annotated SNPs {.tabset .tabset-fade .tabset-pills}

<!-- #### Full anno -->

```{r include = T, eval = T, fig.cap = "Annotation with ChIPseeker, delta"}
delta.meqtl.snp.anno.df <- as.data.frame(delta.meqtl.snp.anno.rds@anno, row.names = names(delta.meqtl.snp.anno.rds@anno))
# delta.meqtl.snp.anno.df$Relation_to_Island <- factor(delta.meqtl.cpg.anno.df$Relation_to_Island, 
#                                                      levels = c("N_Shelf", "N_Shore", "Island", "S_Shore", "S_Shelf", "OpenSea"))
head(delta.meqtl.snp.anno.df, 10)
```

<!-- #### Closest genes -->

<!-- ```{r include = T, eval = T} -->
<!-- head(cpg.closest.genes.dist.df, 100) -->
<!-- ``` -->

<!-- ### Genomic regions -->

<!-- ```{r out.width = "90%", fig.cap = "Distribution of locations of the CpGs in different genomic regions based on Illumina EPIC annotation", include = T, eval = T} -->
<!-- cbPalette <- c(  "#009E73", "#E69F00", "#F0E442", "#D55E00", "#CC79A7", "#56B4E9", "#999999") -->

<!-- ggplot(meqtl.cpg.anno.df, aes(x = Model, fill = Relation_to_Island)) +  -->
<!--   geom_bar( position = "fill") +  -->
<!--   geom_text(aes(by = Model, # scales::percent(..count../tapply(..count.., ..x.., sum), accuracy = 0.1),  -->
<!--                 y = (..count..)/sum(..count..)),  -->
<!--             stat = "prop",  -->
<!--             position = position_fill(vjust = .5), size = 3) + -->
<!--   scale_y_continuous(labels = scales::percent) + -->
<!--   labs(x = "", -->
<!--        y = "Percentage of CpGs",  -->
<!--        title = "") + -->
<!--   theme(legend.position = "right", -->
<!--         panel.grid.major = element_blank(), -->
<!--         panel.background = element_blank(), -->
<!--         plot.title = element_text(size = 8), -->
<!--         axis.title = element_text(size = 8), -->
<!--         axis.text.x = element_text(angle = 0, hjust = 0.5)) + -->
<!--    scale_fill_manual("Type", values = cbPalette) -->
<!-- ``` -->

### Gene regions

```{r include = F, eval = T}
delta.meqtl.snp.anno.stat.df <- delta.meqtl.snp.anno.rds@annoStat
veh.meqtl.snp.anno.stat.df   <- veh.meqtl.snp.anno.rds@annoStat
dex.meqtl.snp.anno.stat.df   <- dex.meqtl.snp.anno.rds@annoStat

delta.meqtl.snp.anno.stat.df[["Model"]] <- as.factor("Delta")
veh.meqtl.snp.anno.stat.df[["Model"]]   <- as.factor("Baseline")
dex.meqtl.snp.anno.stat.df[["Model"]]   <- as.factor("Dex")

meqtl.snp.anno.stat.df <- rbind(delta.meqtl.snp.anno.stat.df,
                                veh.meqtl.snp.anno.stat.df,
                                dex.meqtl.snp.anno.stat.df)

meqtl.snp.anno.stat.df$Feature   <- factor(meqtl.snp.anno.stat.df$Feature)
meqtl.snp.anno.stat.df$Frequency <- signif(meqtl.snp.anno.stat.df$Frequency, 3)
```

```{r out.width = "90%", fig.cap = "Distribution of gene-centric locations of the SNPs based on Illumina EPIC annotation", include = T, eval = T}
cbPalette <- c("#88CCEE", "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888")

ggplot(meqtl.snp.anno.stat.df, aes(x = Model, y = Frequency, fill = Feature)) + 
  geom_bar( position = "fill", stat = "identity") + 
  geom_text(data = subset(meqtl.snp.anno.stat.df, Frequency > 2),
            aes(by = Model, label = Frequency), 
            position = position_fill(vjust = .5), size = 3) +
  labs(x = "",
       y = "Percentage of SNPs", 
       title = "meSNP annotation from UCSC for the hg19 genome build using TxDb.Hsapiens.UCSC.hg19.knownGene and ChIPseeker Bioconductor R packages") +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 0, hjust = 0.5)) +
   scale_fill_manual("Feature", values = cbPalette)
```



<!-- ```{r include = F, eval = T} -->
<!-- library(annotatr) -->
<!-- library(AnnotationHub) -->
<!-- library(TxDb.Hsapiens.UCSC.hg19.knownGene) -->

<!-- annots <- c('hg19_cpgs', 'hg19_basicgenes') #, 'hg19_genes_intergenic', 'hg19_genes_intronexonboundaries') -->
<!-- gene.annots <- c('hg19_basicgenes') #, 'hg19_genes_intergenic', 'hg19_genes_intronexonboundaries') -->

<!-- annotations <- build_annotations( -->
<!--   genome = "hg19", -->
<!--   annotations = gene.annots -->
<!-- ) -->
<!-- ``` -->

<!-- ```{r include = F, eval = T} -->
<!-- delta.meqtls.cpg.annotated <- annotate_regions( -->
<!--     regions = cpg.coord.range, -->
<!--     annotations = annotations, -->
<!--     ignore.strand = TRUE, -->
<!--     quiet = FALSE) -->

<!-- # print(delta.meqtls.cpg.annotated) -->
<!-- # plot_annotation(annotated_regions = delta.meqtls.cpg.annotated) -->

<!-- delta.meqtls.cpg.annotated.df <- data.frame(delta.meqtls.cpg.annotated) %>% setDT() -->

<!-- delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_1to5kb", annot.type := "1-5kb"] -->
<!-- delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_3UTRs", annot.type := "3UTRs"] -->
<!-- delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_5UTRs", annot.type := "5UTRs"] -->
<!-- delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_exons", annot.type := "exons"] -->
<!-- delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_introns", annot.type := "introns"] -->
<!-- delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_promoters", annot.type := "promoters"] -->
<!-- # head(delta.meqtls.cpg.annotated.df) -->

<!-- delta.meqtls.cpg.annotated.df <- delta.meqtls.cpg.annotated.df[, .(seqnames, start, end, annot.type)] %>% unique() -->

<!-- delta.meqtls.cpg.annotated.df$annot.type <- factor(delta.meqtls.cpg.annotated.df$annot.type,  -->
<!--                                                               levels = c("1-5kb", "promoters", "5UTRs", "introns", "exons", "3UTRs")) -->
<!-- ``` -->

<!-- ```{r out.width = "90%", fig.cap = "The distribution of CpGs in different gene regions", include = T, eval = T} -->
<!-- ggplot(delta.meqtls.cpg.annotated.df, aes(x = annot.type)) +  -->
<!--   geom_bar(aes(y = (..count..)/sum(..count..), fill = annot.type), position = position_dodge()) +  -->
<!--   geom_text(aes(label = scales::percent((..count..)/sum(..count..), accuracy = 0.1), y = (..count..)/sum(..count..)),  -->
<!--             stat = "count", vjust = -.5,  position = position_dodge(1), size = 3) + -->
<!--   scale_y_continuous(labels = scales::percent) + -->
<!--   labs(x = "Gene Region", -->
<!--        y = "Percentage of CpG",  -->
<!--        title = "From UCSC Genome Browser") + -->
<!--   theme(legend.title = element_blank(),  -->
<!--         legend.position = "none", #c(.1,.8), -->
<!--         panel.grid.major = element_blank(), -->
<!--         panel.background = element_blank(), -->
<!--         plot.title = element_text(size = 8), -->
<!--         axis.title = element_text(size = 8), -->
<!--         axis.text.x = element_text(angle = 45, hjust = 0.5)) + -->
<!--    scale_fill_manual(values = cbPalette) -->
<!-- ``` -->

## Chromatin States {.tabset .tabset-fade .tabset-pills}

### Annotation table

```{r include = F, eval = T}
# chromhmm.all.states <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/chromHMM_all_states.Rds")
# colnames(values(chromhmm.all.states))[1] <- "type"

chromhmm.states     <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/chromhmm_states.Rds") %>% setDT()
chromhmm.states[COLOR.NAME == "Silver", COLOR.NAME := "Grey"] 
chromhmm.states[COLOR.NAME == "White", COLOR.NAME := "Snow"] 

cbPalette <- chromhmm.states$COLOR.NAME

chromhmm.epigenomes <- read.csv("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/epigenomes.tsv", sep = "\t")
```


```{r func-plot-chromhmm-states, include = F, eval = T}
GetChromHMMPlotAll <- function(df, cbPalette = NULL, plot.title = NULL){
  
  df$annot.type <- factor(df$annot.type, levels = chromhmm.states$STATE)
  df$Model      <- factor(df$Model, levels = c("Delta", "Baseline", "Dex"))
  
  ggplot(df, aes(x = Model, fill = annot.type)) + 
    geom_bar( position = "fill") + 
    # coord_flip() +
    # scale_y_reverse() +
    geom_text(
      aes(by = Model), 
      stat = "prop", 
      position = position_fill(vjust = .5), size = 3) +
    labs(x = "",
       y = "Percentage of SNPs", 
       title = plot.title) +
    theme(legend.position = "right", 
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 8),
          axis.title = element_text(size = 8)) +
    scale_fill_manual("", values = cbPalette) 
}
```

```{r read-chromhmm-anno-delta, eval = T, include = F}
delta.meqtls.snp.chromhmm.annotated.df <- fread(paste0(out.dir.pre, "meqtls_snp_chromhmm_annotated_delta.csv"))

delta.meqtls.snp.chromhmm.annotated.df[, Model := "Delta"]
```

```{r read-chromhmm-anno-veh, eval = T, include = F}
veh.meqtls.snp.chromhmm.annotated.df <- fread(paste0(out.dir.pre, "meqtls_snp_chromhmm_annotated_veh.csv"))

veh.meqtls.snp.chromhmm.annotated.df[, Model := "Baseline"]
```

```{r read-chromhmm-anno-dex, eval = T, include = F}
dex.meqtls.snp.chromhmm.annotated.df <- fread(paste0(out.dir.pre, "meqtls_snp_chromhmm_annotated_dex.csv"))

dex.meqtls.snp.chromhmm.annotated.df[, Model := "Dex"]
```

```{r eval = T, include = F}
meqtls.snp.chromhmm.annotated.plt.df <- rbind(delta.meqtls.snp.chromhmm.annotated.df, 
                                              veh.meqtls.snp.chromhmm.annotated.df,
                                              dex.meqtls.snp.chromhmm.annotated.df)
```

```{r include = T, eval = T}
# meqtls.cpg.chromhmm.annotated <- readRDS(paste0(out.dir.pre.pre, "global_independent_snps/meqtls_cpg_chromhmm_annotated_all_models.rds"))

# meqtls.cpg.chromhmm.annotated.df <- data.frame(meqtls.cpg.chromhmm.annotated) %>% setDT()
# meqtls.cpg.chromhmm.annotated.df[["CpG_ID"]] <- names(meqtls.cpg.chromhmm.annotated)

head(meqtls.snp.chromhmm.annotated.plt.df %>% dplyr::select(SNP, everything()), 100) 
# plot_annotation(annotated_regions = delta.meqtls.cpg.chromhmm.annotated)
```

### All cell types

```{r out.width = "90%", fig.cap = "The distribution of CpGs in 15 chromatin states from 127 cell types", include = T, eval = T}

df <- meqtls.snp.chromhmm.annotated.plt.df[, .(SNP, annot.name, Model)] %>% unique()

plot.title <- "15 chromatin states from 127 cell types"

# GetChromHMMPlotAll(df, cbPalette, plot.title)
# 
df$annot.name <- factor(df$annot.name, levels = chromhmm.states$STATE)
df$Model      <- factor(df$Model, levels = c("Delta", "Baseline", "Dex"))

df.plt <- melt(table(df[, .(annot.name, Model)])) %>% 
   group_by(annot.name, Model, value) %>%
   group_by(Model) %>%
   mutate(Prop = round(value / sum(value) * 100, 2))
 
ggplot(df.plt, aes(x = Model, y = Prop, fill = annot.name )) + 
    geom_bar( position = "fill",  stat = "identity") + 
    geom_text(data = subset(df.plt, Prop > 1.5),
        aes(label = Prop),
        position = position_fill(vjust = .5), size = 3) +
    labs(x = "",
       y = "Percentage of SNPs", 
       title = plot.title) +
    theme(legend.position = "right", 
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 8),
          axis.title = element_text(size = 8)) +
    scale_fill_manual("", values = cbPalette) 
```

### Blood and T-cell

```{r out.width = "90%", fig.cap = "The distribution of CpGs in 15 chromatin states from 14 blood and T-cell types", include = T, eval = T}
chromhmm.blood.eids <- chromhmm.epigenomes[chromhmm.epigenomes$GROUP == "Blood & T-cell", "EID"]

df <- meqtls.snp.chromhmm.annotated.plt.df[annot.code %in% chromhmm.blood.eids, .(SNP, annot.name, Model)] %>% unique()

plot.title <- "15 chromatin states from 14 blood and T-cell types"

df$annot.name <- factor(df$annot.name, levels = chromhmm.states$STATE)
df$Model      <- factor(df$Model, levels = c("Delta", "Baseline", "Dex"))

df.plt <- melt(table(df[, .(annot.name, Model)])) %>% 
   group_by(annot.name, Model, value) %>%
   group_by(Model) %>%
   mutate(Prop = round(value / sum(value) * 100, 2))
 
ggplot(df.plt, aes(x = Model, y = Prop, fill = annot.name )) + 
    geom_bar( position = "fill",  stat = "identity") + 
    geom_text(data = subset(df.plt, Prop > 1.5),
        aes(label = Prop),
        position = position_fill(vjust = .5), size = 3) +
    labs(x = "",
       y = "Percentage of SNPs", 
       title = plot.title) +
    theme(legend.position = "right", 
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 8),
          axis.title = element_text(size = 8)) +
    scale_fill_manual("", values = cbPalette) 

# GetChromHMMPlotAll(df, cbPalette, plot.title)
```

### Brain

```{r out.width = "90%", fig.cap = "The distribution of CpGs in 15 chromatin states from 10 brain types", include = T, eval = T}
chromhmm.brain.eids <- chromhmm.epigenomes[chromhmm.epigenomes$GROUP == "Brain", "EID"]

df <- meqtls.snp.chromhmm.annotated.plt.df[annot.code %in% chromhmm.brain.eids, .(SNP, annot.name, Model)] %>% unique()

plot.title <- "15 chromatin states from 10 brain types"

 df$annot.name <- factor(df$annot.name, levels = chromhmm.states$STATE)
 df$Model      <- factor(df$Model, levels = c("Delta", "Baseline", "Dex"))

 df.plt <- melt(table(df[, .(annot.name, Model)])) %>% 
   group_by(annot.name, Model, value) %>%
   group_by(Model) %>%
   mutate(Prop = round(value / sum(value) * 100, 1.5))
 
ggplot(df.plt, aes(x = Model, y = Prop, fill = annot.name )) + 
    geom_bar( position = "fill",  stat = "identity") + 
    geom_text(data = subset(df.plt, Prop > 1),
        aes(label = Prop),
        position = position_fill(vjust = .5), size = 3) +
    labs(x = "",
       y = "Percentage of SNPs", 
       title = plot.title) +
    theme(legend.position = "right", 
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 8),
          axis.title = element_text(size = 8)) +
    scale_fill_manual("", values = cbPalette) 

# GetChromHMMPlotAll(df, cbPalette, plot.title)
```

# __Functional SNP enrichment__
***

## Chromatin States
<!-- {.tabset .tabset-fade .tabset-pills} -->

```{r eval = T, include = F}
chromhmm.blood.states <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/chromHMM_blood_states.Rds") chromhmm.brain.states <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/chromHMM_brain_states.Rds") 
states.lst <- elementMetadata(chromhmm.blood.states)[, "type"] %>% unique() %>% sort()

epigenome <- fread("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/epigenomes.tsv",
                   col.names = c("EID", "ORDER", "GROUP", "MNEMONIC", "NAME"))
```

```{r eval = T, include = F}
out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/"
chromhmm.blood.enrich.perm.rslt <- read.csv2(paste0(out.dir.pre, "meqtl_snps_chromHMM_blood_enrichment.csv"))

chromhmm.blood.enrich.perm.rslt$state <- factor(chromhmm.blood.enrich.perm.rslt$state, levels = chromhmm.states$STATE)
# levels( chromhmm.blood.enrich.perm.rslt$state) <- chromhmm.states$DESCRIPTION
chromhmm.blood.enrich.perm.rslt$state <- sub("[0-9]*_", "", chromhmm.blood.enrich.perm.rslt$state) 
```

### Permutation results

```{r eval = T, include = T}
rbind(chromhmm.blood.enrich.perm.rslt, chromhmm.brain.enrich.perm.rslt)
```

```{r}
PlotChromHMMEnrichment <- function(enrichment.rslt, plt.title, plt.txt.size = 20){
  enrichment.rslt$state <- factor(enrichment.rslt$state, levels = chromhmm.states$STATE)
  levels(enrichment.rslt$state) <- sub("[0-9]*_", "", enrichment.rslt$state)
  
  enrichment.rslt["is_sign"] <- ifelse(enrichment.rslt$p_val_emp < 0.05, "Significant", "Non-significant")
  
  ggplot(enrichment.rslt, aes(x = state, y = odds.ratio, fill = is_sign)) + 
    geom_bar( stat = "identity", position = position_dodge()) +
    geom_hline(yintercept = 1, color = "red") + 
    labs(x = "",
         y = "Odds ratio", 
         title = plt.title) +
    theme(legend.position = "bottom",
          legend.text = element_text(size = plt.txt.size, colour = "black"),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 0.75 * plt.txt.size),
          axis.title = element_text(size = plt.txt.size),
          axis.text.x = element_text(size = 0.75 * plt.txt.size, angle = 90, colour = "black", hjust = 0.95, vjust = 0.2),
          axis.text.y = element_text(size = 0.75 * plt.txt.size, angle = 0, colour = "black")) +
    scale_fill_manual("", values = c( "#999999", "#009E73")) 
}
```


### Blood

```{r}
tissue <- "blood"

enrichment.rslts.fn <- paste0(out.dir.pre, "meqtl_snps_chromHMM_", tissue, "/", list.files(paste0(out.dir.pre, "meqtl_snps_chromHMM_", tissue)))

doc <- read_pptx("~/bio/code/mpip/dex-stim-human-array/output/plots/enrichment/meqtl_snp_chromHMM_enrichment.pptx")

enrichment.all.df <- lapply(seq_along(enrichment.rslts.fn), function(i) {
  eid   <-  sub(".csv", "", sub(".*_", "", enrichment.rslts.fn[i]))
  ename <- epigenome[EID == eid, NAME] 
  plot.title <- paste0("Histone mark enrichment for the delta meSNPs for ", ename)
  
  df  <- read.csv2(enrichment.rslts.fn[i])
  ggp <- PlotChromHMMEnrichment(df, plt.title = plot.title)
  
  add_slide(doc,
            layout = "Plot",
            master = "Basic") %>%
    ph_with(
      value = ggp,
      location = ph_location_label(ph_label = "Content Placeholder 2")
    )
  return(data.frame(df, ename = ename))
  }) %>%
  bind_rows()

enrichment.all.df$state <- factor(enrichment.all.df$state, levels = chromhmm.states$STATE)
levels(enrichment.all.df$state) <- sub("[0-9]*_", "", enrichment.all.df$state)

plt.txt.size <- 20
ggp.heatmap <- 
  ggplot(enrichment.all.df, aes(state, ename, fill = p_val_emp)) + 
    geom_tile(colour = "white", size = 0.25) +
    theme(legend.position = "right",
          legend.title = element_text(size = plt.txt.size, colour = "black"),
          legend.text = element_text(size = plt.txt.size, colour = "black"),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 0.75 * plt.txt.size),
          axis.title = element_blank(),
          axis.text.x = element_text(size = 0.75 * plt.txt.size, angle = 90, colour = "black", hjust = 0.95, vjust = 0.2),
          axis.text.y = element_text(size = 0.75 * plt.txt.size, angle = 0, colour = "black")) +
  scale_fill_viridis()

add_slide(doc,
            layout = "Plot",
            master = "Basic") %>%
    ph_with(
      value = ggp.heatmap,
      location = ph_location_label(ph_label = "Content Placeholder 2")
    )

print(doc, target = paste0("~/bio/code/mpip/dex-stim-human-array/output/plots/enrichment/meqtl_snp_chromHMM_", tissue, "_enrichment.pptx"))
```


### Brain

```{r}
tissue <- "brain"

enrichment.rslts.fn <- paste0(out.dir.pre, "meqtl_snps_chromHMM_", tissue, "/", list.files(paste0(out.dir.pre, "meqtl_snps_chromHMM_", tissue)))
enrichmnet.rslts.df <- lapply(enrichment.rslts.fn, read.csv2)

doc <- read_pptx("~/bio/code/mpip/dex-stim-human-array/output/plots/enrichment/meqtl_snp_chromHMM_enrichment.pptx")

lapply(seq_along(enrichmnet.rslts.df), function(i) {
  eid   <-  sub(".csv", "", sub(".*_", "", enrichment.rslts.fn[i]))
  ename <- epigenome[EID == eid, NAME] 
  plot.title <- paste0("Histone mark enrichment for the delta meSNPs for ", ename)
  ggp <- PlotChromHMMEnrichment (enrichmnet.rslts.df[[i]], 
                                 plt.title = plot.title)
  
  add_slide(doc,
            layout = "Plot",
            master = "Basic") %>%
    ph_with(
      value = ggp,
      location = ph_location_label(ph_label = "Content Placeholder 2")
    )
  })

print(doc, target = paste0("~/bio/code/mpip/dex-stim-human-array/output/plots/enrichment/meqtl_snp_chromHMM_", tissue, "_enrichment.pptx"))
```

#### Delta vs Baseline

```{r out.width = "90%", fig.cap = "Histone mark enrichment for the delta meSNPs.  The Y-axis denotes the fold enrichment/depletion as compared to baseline meSNPs. Green bars indicate significant enrichment/depletion, grey bars non-significant differences based on permutation Fisher-tests empirical P <= 0.05. ", eval = T, include = T}
PlotChromHMMEnrichment(chromhmm.blood.enrich.perm.rslt, plt.title = "Histone mark enrichment for the delta meSNPs, delta vs baseline")
```

```{r out.width = "90%", fig.cap = "Histone mark enrichment for the delta meSNPs. The Y-axis denotes the proportion of overlaps between meSNPs and blood cell type states.", eval = T, include = T}

chromhmm.blood.enrich.perm.rslt["prop"] <-  signif(chromhmm.blood.enrich.perm.rslt$n_snps_overlap / delta.meqtl.snp.anno.rds@peakNum * 100, 2)

ggplot(chromhmm.blood.enrich.perm.rslt, aes(x = state, y = prop)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
  # geom_text(aes(label = paste0(n_snps_overlap, " \n(", prop, "%)")), position = position_dodge(width = 0.9), vjust = -0.25, cex = 2) + 
  geom_text(aes(label = n_snps_overlap), position = position_dodge(width = 0.9), vjust = -0.25, cex = 5) + 
  labs(x = "",
       y = "Overlaps, %", 
       title = "") +
  theme(legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 20),
        axis.text.x = element_text(size = 0.75 * 20, angle = 90, hjust = 1, color = "black"),
        axis.text.y = element_text(size = 0.75 * 20, angle = 90, color = "black")) +
   scale_fill_manual("", values = c( "#009E73")) 
```

#### Delta vs All

```{r eval = T, include = F}
out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/"
chromhmm.blood.enrich.perm.rslt <- read.csv2(paste0(out.dir.pre, "meqtl_snps_chromHMM_blood_enrichment_delta_vs_all.csv"))

chromhmm.blood.enrich.perm.rslt$state <- factor(chromhmm.blood.enrich.perm.rslt$state, levels = chromhmm.states$STATE)
levels( chromhmm.blood.enrich.perm.rslt$state) <- chromhmm.states$DESCRIPTION
# chromhmm.blood.enrich.perm.rslt$state <- sub("[0-9]*_", "", chromhmm.blood.enrich.perm.rslt$state) 
```

```{r out.width = "90%", fig.cap = "Histone mark enrichment for the delta meSNPs.  The Y-axis denotes the fold enrichment/depletion of delta meSNPs as compared to all SNPs. Green bars indicate significant enrichment/depletion, grey bars non-significant differences based on permutation Fisher-tests empirical P <= 0.05. ", eval = T, include = T}

chromhmm.blood.enrich.perm.rslt["is_sign"] <- ifelse(chromhmm.blood.enrich.perm.rslt$p_val_emp < 0.05, "Significant", "Non-significant")

ggplot(chromhmm.blood.enrich.perm.rslt, aes(x = state, y = odds.ratio, fill = is_sign)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
  geom_hline(yintercept = 1, color = "red") + 
  labs(x = "",
       y = "Odds ratio", 
       title = "") +
  theme(legend.position = "bottom",
        legend.text = element_text(size = 20, colour = "black"),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 20),
        axis.text.x = element_text(size = 15, angle = 90, colour = "black"),
        axis.text.y = element_text(size = 16, angle = 0, colour = "black")) +
   scale_fill_manual("", values = c( "#999999", "#009E73")) 
```

#### Baseline vs All

```{r eval = T, include = F}
out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/"
chromhmm.blood.enrich.perm.rslt <- read.csv2(paste0(out.dir.pre, "meqtl_snps_chromHMM_blood_enrichment_veh_vs_all.csv"))

chromhmm.blood.enrich.perm.rslt$state <- factor(chromhmm.blood.enrich.perm.rslt$state, levels = chromhmm.states$STATE)
levels( chromhmm.blood.enrich.perm.rslt$state) <- chromhmm.states$DESCRIPTION
# chromhmm.blood.enrich.perm.rslt$state <- sub("[0-9]*_", "", chromhmm.blood.enrich.perm.rslt$state) 
```

```{r out.width = "90%", fig.cap = "Histone mark enrichment for the delta meSNPs.  The Y-axis denotes the fold enrichment/depletion of baseline meSNPs as compared to all SNPs. Green bars indicate significant enrichment/depletion, grey bars non-significant differences based on permutation Fisher-tests empirical P <= 0.05. ", eval = T, include = T}

chromhmm.blood.enrich.perm.rslt["is_sign"] <- ifelse(chromhmm.blood.enrich.perm.rslt$p_val_emp < 0.05, "Significant", "Non-significant")

ggplot(chromhmm.blood.enrich.perm.rslt, aes(x = state, y = odds.ratio, fill = is_sign)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
  geom_hline(yintercept = 1, color = "red") + 
  labs(x = "",
       y = "Odds ratio", 
       title = "") +
  theme(legend.position = "bottom",
        legend.text = element_text(size = 20, colour = "black"),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 20),
        axis.text.x = element_text(size = 15, angle = 90, colour = "black"),
        axis.text.y = element_text(size = 16, angle = 0, colour = "black")) +
   scale_fill_manual("", values = c( "#999999", "#009E73")) 
```

### Brain

#### Delta vs Baseline

```{r out.width = "90%", fig.cap = "Histone mark enrichment for the delta meSNPs.  The Y-axis denotes the fold enrichment/depletion as compared to baseline meSNPs. Green bars indicate significant enrichment/depletion, grey bars non-significant differences based on permutation Fisher-tests empirical P <= 0.05. ", eval = T, include = T}
chromhmm.brain.enrich.perm.rslt <- read.csv2(paste0(out.dir.pre, "meqtl_snps_chromHMM_brain_enrichment.csv"))
PlotChromHMMEnrichment(chromhmm.brain.enrich.perm.rslt, plt.title = "Histone mark enrichment for the delta meSNPs, delta vs baseline")
```

```{r out.width = "90%", fig.cap = "Histone mark enrichment for the delta meSNPs. The Y-axis denotes the proportion of overlaps between meSNPs and brain cell type states.", eval = T, include = T}

chromhmm.brain.enrich.perm.rslt$state <- factor(chromhmm.brain.enrich.perm.rslt$state, levels = chromhmm.states$STATE)
chromhmm.brain.enrich.perm.rslt$state <- sub("[0-9]*_", "", chromhmm.brain.enrich.perm.rslt$state)

chromhmm.brain.enrich.perm.rslt["prop"] <- signif(chromhmm.brain.enrich.perm.rslt$n_snps_overlap / delta.meqtl.snp.anno.rds@peakNum * 100, 2)

ggplot(chromhmm.brain.enrich.perm.rslt, aes(x = state, y = prop)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
# geom_text(aes(label = paste0(n_snps_overlap, " (", prop, "%)")), position = position_dodge(width = 0.9), vjust = -0.25, cex = 3) + 
  geom_text(aes(label = n_snps_overlap), position = position_dodge(width = 0.9), vjust = -0.25, cex = 5) + 
  labs(x = "",
       y = "Overlaps, %", 
       title = "") +
  theme(legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 20),
        axis.text.x = element_text(size = 0.75 * 20, angle = 90, hjust = 1, color = "black"),
        axis.text.y = element_text(size = 0.75 * 20, angle = 90, color = "black")) +
   scale_fill_manual("", values = c( "#009E73")) 
```

#### Delta vs All

```{r eval = T, include = F}
out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/"
chromhmm.brain.enrich.perm.rslt <- read.csv2(paste0(out.dir.pre, "meqtl_snps_chromHMM_brain_enrichment_delta_vs_all.csv"))

PlotChromHMMEnrichment(chromhmm.brain.enrich.perm.rslt, plt.title = "Histone mark enrichment for the delta meSNPs, delta vs all")
```

#### Baseline vs All

```{r eval = T, include = F}
out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/"
chromhmm.brain.enrich.perm.rslt <- read.csv2(paste0(out.dir.pre, "meqtl_snps_chromHMM_brain_enrichment_veh_vs_all.csv"))

PlotChromHMMEnrichment(chromhmm.brain.enrich.perm.rslt, plt.title = "Histone mark enrichment for the delta meSNPs, baseline vs all")
```

## GWAS Enrichment

```{r eval = T, include = F}
out.dir.pre  <- "~/bio/code//mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/"
gwas.enrich.perm.rslt <- read.csv2(paste0(out.dir.pre, "meqtl_snps_GWAS_enrichment_perm_1000.csv"))

# gwas.enrich.perm.rslt ["data"] <- c("Nominal: p <= 0.05", "Genome-wide: p <= 5e-8")

gwas.enrich.perm.rslt$odds.ratio <- as.numeric(gwas.enrich.perm.rslt$odds.ratio)
gwas.enrich.perm.rslt$p_val      <- as.numeric(gwas.enrich.perm.rslt$p_val)
gwas.enrich.perm.rslt$p_val_emp  <- as.numeric(gwas.enrich.perm.rslt$p_val_emp)

gwas.enrich.perm.rslt$data   <- factor(gwas.enrich.perm.rslt$data)
gwas.enrich.perm.rslt$data <- c("Cross Disorders", "ADHD", "ASD", "BP", "MDD", "IBD", "BMI", "SCZ", "PD")
```


```{r out.width = "90%", fig.cap = "Enrichment for nominal and genome-wide significant GWAS associations for the delta meSNPs for CrossDisorder GWAS. The Y-axis denotes the fold enrichment with regard to baseline meSNPs. Green bars indicate significant enrichment/depletion, grey bars non-significant differences based on permutation Fisher-tests empirical P <= 0.05.", eval = T, include = T}

gwas.enrich.perm.rslt["is_sign"] <- ifelse(gwas.enrich.perm.rslt$p_val_emp < 0.05, "Significant", "Non-significant")

ggplot(gwas.enrich.perm.rslt, aes(x = data, y = odds.ratio, fill = is_sign)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
  geom_hline(yintercept = 1, color = "red") + 
  labs(x = "",
       y = "Odds ratio", 
       title = "") +
  theme(legend.position = "bottom",
        legend.text = element_text(size = 20, colour = "black"),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 20),
        axis.text.x = element_text(size = 20, angle = 0, colour = "black"),
        axis.text.y = element_text(size = 16, angle = 0, colour = "black")) +
   scale_fill_manual("", values = c( "#999999", "#009E73")) 
```

```{r out.width = "90%", fig.cap = "Enrichment of the delta meSNPs for different GWAS. The Y-axis denotes the proportion of overlaps between meSNPs and GWAS SNPs.", eval = T, include = T}

gwas.enrich.perm.rslt["prop"] <-  signif(gwas.enrich.perm.rslt$n_snps_overlap / delta.meqtl.snp.anno.rds@peakNum * 100, 2)

ggplot(gwas.enrich.perm.rslt, aes(x = data, y = prop)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
# geom_text(aes(label = paste0(n_snps_overlap, "\n(", prop, "%)")), position = position_dodge(width = 0.9), vjust = -0.25, cex = 3) + 
  geom_text(aes(label = n_snps_overlap), position = position_dodge(width = 0.9), vjust = -0.25, cex = 3) + 
  labs(x = "",
       y = "Overlaps, %", 
       title = "") +
  theme(legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 45, hjust = 1)) +
   scale_fill_manual("", values = c( "#009E73")) 
```
